code/datums/components/customizable_reagent_holder.dm

```diff
@@ -0,0 +1,245 @@
+	///List of item ingredients.
+	var/list/obj/item/ingredients
+	///Type path of replacement atom.
+	var/replacement
+	///Type of fill, can be [CUSTOM_INGREDIENT_ICON_NOCHANGE] for example.
+	var/fill_type
+	///Number of max ingredients.
+	var/max_ingredients
+	///Overlay used for certain fill types, always shows up on top.
+	var/mutable_appearance/top_overlay
+	///Type of ingredients to accept, [CUSTOM_INGREDIENT_TYPE_EDIBLE] for example.
+	var/ingredient_type
+
+
+/datum/component/customizable_reagent_holder/Initialize(
+		atom/replacement,
+		fill_type,
+		ingredient_type = CUSTOM_INGREDIENT_TYPE_EDIBLE,
+		max_ingredients = INFINITY,
+		list/obj/item/initial_ingredients = null)
+	if(!isatom(parent))
+		return COMPONENT_INCOMPATIBLE
+	var/atom/P = parent
+	if (!P.reagents)
+		return COMPONENT_INCOMPATIBLE
+
+	src.replacement = replacement
+	src.fill_type = fill_type
+	src.max_ingredients = max_ingredients
+	src.ingredient_type = ingredient_type
+
+	if (initial_ingredients)
+		for (var/ingr in initial_ingredients)
+			var/obj/item/I = ingr
+			add_ingredient(I)
+			handle_fill(I)
+
+
+/datum/component/customizable_reagent_holder/Destroy(force, silent)
+	QDEL_NULL(top_overlay)
+	return ..()
+
+
+/datum/component/customizable_reagent_holder/RegisterWithParent()
+	. = ..()
+	RegisterSignal(parent, COMSIG_PARENT_ATTACKBY, .proc/customizable_attack)
+	RegisterSignal(parent, COMSIG_PARENT_EXAMINE, .proc/on_examine)
+	RegisterSignal(parent, COMSIG_ATOM_PROCESSED, .proc/on_processed)
+
+
+/datum/component/customizable_reagent_holder/UnregisterFromParent()
+	. = ..()
+	UnregisterSignal(parent, list(
+			COMSIG_PARENT_ATTACKBY,
+			COMSIG_PARENT_EXAMINE,
+			COMSIG_ATOM_PROCESSED
+		)
+	)
+
+
+/datum/component/customizable_reagent_holder/PostTransfer()
+	if(!isatom(parent))
+		return COMPONENT_INCOMPATIBLE
+	var/atom/P = parent
+	if (!P.reagents)
+		return COMPONENT_INCOMPATIBLE
+
+///Handles when the customizable food is examined.
+/datum/component/customizable_reagent_holder/proc/on_examine(atom/A, mob/user, list/examine_list)
+	SIGNAL_HANDLER
+
+	var/atom/P = parent
+	var/ingredients_listed = ""
+	if (LAZYLEN(ingredients))
+		for (var/i in 1 to ingredients.len)
+			var/obj/item/I = ingredients[i]
+			var/ending = ", "
+			switch(length(ingredients))
+				if (2)
+					if (i == 1)
+						ending = " and "
+				if (3 to INFINITY)
+					if (i == ingredients.len - 1)
+						ending = ", and "
+			ingredients_listed += "\a [I.name][ending]"
+	examine_list += "It contains [LAZYLEN(ingredients) ? "[ingredients_listed]" : " no ingredients, "]making a [custom_adjective()]-sized [initial(P.name)]."
+
+
+///Handles when the customizable food is attacked by something.
+/datum/component/customizable_reagent_holder/proc/customizable_attack(datum/source, obj/item/I, mob/M, silent = FALSE, force = FALSE)
+	SIGNAL_HANDLER
+
+	var/valid_ingredient = TRUE
+
+	switch (ingredient_type)
+		if (CUSTOM_INGREDIENT_TYPE_EDIBLE)
+			valid_ingredient = IS_EDIBLE(I)
```

Comment:
Why a `switch` and not an `if`?

Reply:
because it's an enum and he will most likely add more types later